client/rest/src/main/java/org/elasticsearch/client/RuntimeInfo.java

```diff
@@ -0,0 +1,145 @@
+        String v;
+
+        v = HlrcKind();
+        if (v != null) {
+            s.append(",hl=").append(v);
+        }
+
+        v= kotlinVersion();
+        if (v != null) {
+            s.append(",kt=").append(v);
+        }
+
+        v = scalaVersion();
+        if (v != null) {
+            s.append(",sc=").append(v);
+        }
+
+        v = groovyVersion();
+        if (v != null) {
+            s.append(",clj=").append(v);
+        }
+
+        v = groovyVersion();
+        if (v != null) {
+            s.append(",gy=").append(v);
+        }
+
+        return s.toString();
+    }
+
+    public static String HlrcKind() {
+        try {
+            Class.forName("org.elasticsearch.client.RestHighLevelClient");
+        } catch (Throwable t) {
+            return null;
+        }
+        // 1 is the HLRC based on ES server request/response classes
+        // 2 will be the HLRC based on code generation from API specs
+        return "1";
+    }
+
+    public static String kotlinVersion() {
+        try {
+            //KotlinVersion.CURRENT.toString()
+            Class<?> clazz = Class.forName("kotlin.KotlinVersion");
+            Field field = clazz.getField("CURRENT");
+            String version = field.get(null).toString();
+            return stripPatchRevision(version);
+
+        } catch (Throwable t) {
+            // ignore
+        }
+        return null;
+    }
+
+    public static String scalaVersion() {
+        try {
+            // scala.util.Properties.versionNumberString()
+            Class<?> clazz = Class.forName("scala.util.Properties");
+            Method m = clazz.getMethod("versionNumberString");
+            String version = (String) m.invoke(null);
+            return stripPatchRevision(version);
+
+        } catch (Throwable t) {
+            // ignore
+        }
+        return null;
+    }
+
+    public static String clojureVersion() {
+        try {
+            // (clojure-version) which translates to
+            // clojure.core$clojure_version.invokeStatic()
+            Class<?> clazz = Class.forName("clojure.core$clojure_version");
+            Method m = clazz.getMethod("invokeStatic");
+            String version = (String) m.invoke(null);
+            return stripPatchRevision(version);
+
+        } catch (Throwable t) {
+            // ignore
+        }
+        return null;
+    }
+
+    public static String groovyVersion() {
+        try {
+            // groovy.lang.GroovySystem.getVersion()
+            // There's also getShortVersion(), but only since Groovy 3.0.1
+            Class<?> clazz = Class.forName("groovy.lang.GroovySystem");
+            Method m = clazz.getMethod("getVersion");
+            String version = (String) m.invoke(null);
+            return stripPatchRevision(version);
+
+        } catch (Throwable t) {
+            // ignore
+        }
+        return null;
+    }
+
+    private static String stripPatchRevision(String version) {
```

Comment:
Can we feed this method a large input of version formats in a test to make sure it does the right thing? (I guess it'd be more about looking into the various languages versioning strategies)

Reply:
They all use semantic versioning but have slight differences in how they name non-final releases. I added `RuntimeInfoTests` that checks various version naming strategies for the languages we test (information found on https://search.maven.org/)

Reply:
This is great, thank you!